runtime/src/main/java/org/corfudb/runtime/view/stream/IStreamAddressSpace.java

```diff
@@ -0,0 +1,169 @@
+package org.corfudb.runtime.view.stream;
+
+import java.util.List;
+import java.util.function.Function;
+
+import org.corfudb.protocols.wireprotocol.ILogData;
+import org.corfudb.runtime.view.StreamOptions;
+
+/**
+ * <p>This interface represents the space of addresses of a stream. It also defines
+ * how the space of addresses can be traversed and discovered, i.e.,
+ * how to find the addresses belonging to updates to this stream.</p>
+ *
+ * <p>The space of addresses is tracked by a pointer which allows to move back and forth between
+ * different versions (states) of the stream.</p>
+ *
+ * Created by amartinezman on 4/24/18.
+ */
+public interface IStreamAddressSpace {
+
+    /**
+     * Reset the address space, i.e., reset pointers.
+     *
+     */
+    void reset() ;
+
+    /**
+     * Seeks for a specific address, and moves the pointer in the stream to this position.
+     *
+     */
+    void seek(long address);
+
+    /**
+     * Get maximum resolved address.
+     *
+     * @return maximum address for this stream (address is global)
+     */
+    long getMax();
+
+    /**
+     * Get minimum resolved address.
+     *
+     * @return minimum address for this stream (address is global)
+     */
+    long getMin();
+
+    /**
+     * Returns the next address in the stream
+     *
+     * @return next address in the stream (address is global)
+     */
+    long next();
+
+    /**
+     * Returns the previous address in the stream
+     *
+     * @return previous address in the stream (address is global)
+     */
+    long previous();
+
+    /**
+     * Returns a list of addresses up to a given limit.
+     *
+     * @return addresses in the stream up to a limit (addresses are global)
+     */
+    List<Long> remainingUpTo(long limit);
+
+    /**
+     * Add list of addresses to the space of addresses of this stream.
+     *
+     * @param addresses global addresses to be added.
+     */
+    void addAddresses(List<Long> addresses);
```

Comment:
If you want to use the List<> collection, I would recommend at least annotating this parameter @NonNull.